### PR TITLE
feat(ui5-calendar): add disabled dates functionality

### PR DESCRIPTION
**Pull request by GDamyanov**
This PR introduces support for disabling specific date ranges in the ui5-calendar component via the new disabledDates slot. Applications can now pass one or more ui5-date-range elements to the calendar, specifying [startValue](vscode-file://vscode-app/private/var/folders/85/p7hs9tfx5gvcx9zyf28khlb80000gn/T/AppTranslocation/DC8B5CA5-736C-436A-B978-50C80CCDFBB5/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) and/or [endValue](vscode-file://vscode-app/private/var/folders/85/p7hs9tfx5gvcx9zyf28khlb80000gn/T/AppTranslocation/DC8B5CA5-736C-436A-B978-50C80CCDFBB5/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) to mark dates as non-selectable.

Key Changes

- Added disabledDates slot to ui5-calendar for passing disabled date ranges.

- Implemented logic in DayPicker to prevent selection of dates within disabled ranges.

- Updated internal date parsing to respect the calendar’s [formatPattern](vscode-file://vscode-app/private/var/folders/85/p7hs9tfx5gvcx9zyf28khlb80000gn/T/AppTranslocation/DC8B5CA5-736C-436A-B978-50C80CCDFBB5/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html).

- Added comprehensive JSDoc documentation for new properties and methods.

- Created Cypress tests to verify disabled date functionality, including edge cases and format pattern support.

- Updated sample and test pages to demonstrate usage.

Motivation
This feature allows applications to restrict user selection to valid dates only, improving UX for scenarios like booking, scheduling, or compliance.

### CODE REVIEW

File: packages/main/cypress/specs/Calendar.cy.tsx [hunk 1]
```diff
@@ -934,6 +934,154 @@ describe("Calendar general interaction", () => {
...
+		cy.ui5CalendarGetDay("#calendar1", disabledDate.toString())
+			.realClick();
+
+		// Verify the date was not selected
+		cy.get<Calendar>("#calendar1")
+			.invoke("prop", "selectedDates")
+			.should("have.length", 0);
+	});
+
+	it("Disabled date range with only start date disables dates from start onwards", () => {
```
Comment (Stoev) [inline packages/main/cypress/specs/Calendar.cy.tsx]:
```suggestion
	it("Disables a single date equal to start date when end date is not defined", () => {
```

File: packages/main/cypress/specs/Calendar.cy.tsx [hunk 2]
```diff
@@ -934,6 +934,154 @@ describe("Calendar general interaction", () => {
...
+		cy.ui5CalendarGetDay("#calendar1", startDate.toString())
+			.should("have.class", "ui5-dp-item--disabled");
+
+		// Date after start should not be disabled
+		const afterStartDate = new Date(Date.UTC(2024, 10, 17, 0, 0, 0)).valueOf() / 1000;
+		cy.ui5CalendarGetDay("#calendar1", afterStartDate.toString())
+			.should("not.have.class", "ui5-dp-item--disabled");
+	});
+
+	it("Disabled date range with only end date disables dates up to end", () => {
```
Comment (Stoev) [inline packages/main/cypress/specs/Calendar.cy.tsx]:
```suggestion
	it("Disables all dates before end date when start date is not defined", () => {
```

**Review by Todor-ads — APPROVED** [review]

File: packages/main/test/pages/Calendar.html [hunk 3]
```diff
@@ -178,6 +179,22 @@
...
+	<section>
+		<ui5-title>Calendar with Disabled Dates</ui5-title>
+		<ui5-calendar id="calendar-disabled-dates">
+			<ui5-date-range slot="disabledDates" start-value="Nov 15, 2025" end-value="Nov 20, 2025"></ui5-date-range>
+			<ui5-date-range slot="disabledDates" start-value="Nov 25, 2025" end-value="Nov 27, 2025"></ui5-date-range>
+		</ui5-calendar>
+	</section>
+
+	<section>
+		<ui5-title>Calendar with Disabled Dates with format pattern</ui5-title>
```
Comment (Stoev) [inline packages/main/test/pages/Calendar.html]:
```suggestion
		<ui5-title>Calendar with Disabled Dates using a format pattern</ui5-title>
```

File: packages/main/src/DayPicker.ts [hunk 4]
```diff
@@ -818,6 +825,63 @@ class DayPicker extends CalendarPart implements ICalendarPicker {
...
+	 * Checks if a given date is disabled (not selectable).
+	 * A date is considered disabled if:
+	 * - It falls outside the min/max date range defined by the component
+	 * - It matches a single disabled date
+	 * - It falls within a disabled date range (inclusive of start and end dates)
+	 * @param date - The date to check
+	 * @returns `true` if the date is enabled (selectable), `false` if disabled
+	 * @private
+	 */
+	_isDateEnabled(date: CalendarDate): boolean {
```
Comment (Stoev) [inline packages/main/src/DayPicker.ts:838]: First sentence: "Checks if a given date is disabled (not selectable)." — but the method name and @returns describe the opposite (returns true if enabled).

File: packages/main/src/DayPicker.ts [hunk 5]
```diff
@@ -818,6 +825,63 @@ class DayPicker extends CalendarPart implements ICalendarPicker {
 			|| (iWeekendEnd < iWeekendStart && (iWeekDay >= iWeekendStart || iWeekDay <= iWeekendEnd));
 	}
 
+	/**
+	 * Checks if a given date is disabled (not selectable).
+	 * A date is considered disabled if:
+	 * - It falls outside the min/max date range defined by the component
+	 * - It matches a single disabled date
+	 * - It falls within a disabled date range (inclusive of start and end dates)
```
Comment (Stoev) [inline packages/main/src/DayPicker.ts]: Please double-check the validity of this Copilot suggestion:

Bulleted list states ranges are "inclusive of start and end dates", while the implementation uses exclusive comparisons (">" and "<") when both start and end are present. If you intend inclusivity, update the code to use >= / <= to match the original wording.

File: packages/main/src/DayPicker.ts (hunk 4 above)
Comment (GDamyanov) [reply packages/main/src/DayPicker.ts:838]: Fixed in [555c611](https://github.com/UI5/webcomponents/pull/12691/commits/555c611816a760544ea0658039f6aa817ee836f1)

File: packages/main/src/DayPicker.ts (hunk 5 above)
Comment (GDamyanov) [reply packages/main/src/DayPicker.ts]: Fixed in [555c611](https://github.com/UI5/webcomponents/pull/12691/commits/555c611816a760544ea0658039f6aa817ee836f1)

**Review by Stoev — APPROVED** [review]

File: packages/main/cypress/specs/Calendar.cy.tsx [hunk 6]
```diff
@@ -936,6 +949,159 @@ describe("Calendar general interaction", () => {
...
+		cy.ui5CalendarGetDay("#calendar1", disabledDate.toString())
+			.should("have.class", "ui5-dp-item--disabled")
+			.should("have.attr", "aria-disabled", "true");
+
+		// Try to click on a disabled date
+		cy.ui5CalendarGetDay("#calendar1", disabledDate.toString())
+			.realClick();
+
+		// Verify the date was not selected
+		cy.get<Calendar>("#calendar1")
```
Comment (tsanislavgatev) [inline packages/main/cypress/specs/Calendar.cy.tsx]: why don't we just check if the day that we are clicking on does not have specific class, but we need to invoke a property and check there?

File: packages/main/cypress/specs/Calendar.cy.tsx [hunk 7]
```diff
@@ -936,6 +949,159 @@ describe("Calendar general interaction", () => {
...
+		cy.ui5CalendarGetDay("#calendar1", validStartDate.toString())
+			.realClick();
+
+		// Try to select an end date in the disabled range
+		const disabledEndDate = new Date(Date.UTC(2024, 10, 12, 0, 0, 0)).valueOf() / 1000;
+		cy.ui5CalendarGetDay("#calendar1", disabledEndDate.toString())
+			.realClick();
+
+		// The disabled date should not be selectable
+		cy.get<Calendar>("#calendar1")
```
Comment (tsanislavgatev) [inline packages/main/cypress/specs/Calendar.cy.tsx]: same as prev comment

File: packages/main/src/Calendar.ts [hunk 8]
```diff
@@ -778,6 +806,10 @@ class Calendar extends CalendarPart {
 		return this.getSlottedNodes<SpecialCalendarDate>("specialDates");
 	}
 
+	get _disabledDateRanges() {
```
Comment (tsanislavgatev) [inline packages/main/src/Calendar.ts]: Can't we get this with just this.disabledDates?

File: packages/main/src/DayPicker.ts [hunk 9]
```diff
@@ -818,6 +825,63 @@ class DayPicker extends CalendarPart implements ICalendarPicker {
...
+
+			if (endTimestamp) {
+				return dateTimestamp > startTimestamp && dateTimestamp < endTimestamp;
+			}
+
+			if (startTimestamp && !endTimestamp) {
+				return dateTimestamp === startTimestamp;
+			}
+
+			return false;
```
Comment (tsanislavgatev) [inline packages/main/src/DayPicker.ts]: This can be optimised to:

if (endTimestamp) {
  return dateTimestamp > startTimestamp && dateTimestamp < endTimestamp;
}
return startTimestamp && dateTimestamp === startTimestamp;

File: packages/main/src/DayPicker.ts [hunk 10]
```diff
@@ -818,6 +825,63 @@ class DayPicker extends CalendarPart implements ICalendarPicker {
...
+	 * @param dateValue - Date string to convert
+	 * @returns timestamp in seconds, or 0 if invalid
+	 * @private
+	 */
+	_getTimestampFromDateValue(dateValue?: string): number {
+		if (!dateValue) {
+			return 0;
+		}
+
+		try {
```
Comment (tsanislavgatev) [inline packages/main/src/DayPicker.ts:869]: Are we sure that either of the methods throw an error to be caught?

If not we can just go for:

const jsDate = this.getValueFormat().parse(dateValue);
const calendarDate = CalendarDate.fromLocalJSDate(
  jsDate,
  this._primaryCalendarType,
);
return calendarDate.valueOf() / 1000;

And if needed to check if there's calendarDate before returning it.

File: packages/main/cypress/specs/Calendar.cy.tsx (hunk 6 above)
Comment (GDamyanov) [reply packages/main/cypress/specs/Calendar.cy.tsx]: Fixed in [5decd79](https://github.com/UI5/webcomponents/pull/12691/commits/5decd7995e6afa6b835d2921a12f6b174eebe69d)

File: packages/main/cypress/specs/Calendar.cy.tsx (hunk 7 above)
Comment (GDamyanov) [reply packages/main/cypress/specs/Calendar.cy.tsx]: Fixed in [5decd79](https://github.com/UI5/webcomponents/pull/12691/commits/5decd7995e6afa6b835d2921a12f6b174eebe69d)

File: packages/main/src/Calendar.ts (hunk 8 above)
Comment (GDamyanov) [reply packages/main/src/Calendar.ts]: Fixed in [9d30dc5](https://github.com/UI5/webcomponents/pull/12691/commits/9d30dc515f7774672961b28281aaa2e39b11f6fc)

File: packages/main/src/DayPicker.ts (hunk 9 above)
Comment (GDamyanov) [reply packages/main/src/DayPicker.ts]: Fixed in [e601fb8](https://github.com/UI5/webcomponents/pull/12691/commits/e601fb83ed6c9389cdf37a19d59eb0c91b1f2d9f)

File: packages/main/src/DayPicker.ts (hunk 10 above)
Comment (GDamyanov) [reply packages/main/src/DayPicker.ts:869]: fromLocalJSDate method throws exception.
